### DISCUSSION TITLE
v0.16.0

### DISCUSSION THREAD

**Discussion by nielsvanvelzen** [jellyfin/jellyfin-androidtv#3214] [General]
It has been a busy year with a lot of work on the Android TV app. After an extended beta period today is finally the day to leave it and publish a stable release. Read the [blog post](https://jellyfin.org/posts/androidtv-v0.16.0) to learn about the changes in this release, or read the changelog below.

## 🌟 Highlights

- Show resumable episodes in smart screen for shows #2361, by [@Clubfan22](https://github.com/Clubfan22)
- Show currently playing music on screensaver #2512, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add in\-app screensaver #2539, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add separate buttons for artists and album artists in music smart screen #2667, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Use series poster in latest row when "prefer series thumbnails" is enabled #3060, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)

## 🏗️ Enhancements

- Add AV1 direct play support\. #2392, by [@MWisBest](https://github.com/MWisBest)
- Add playback rewrite backend and plugin systems #2401, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add media management preference #2415, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add ExoPlayer backend to playback rewrite #2427, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add PlaySessionService to playback rewrite #2428, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Update playback module with SDK 1\.4\.1 changes #2441, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add option to enable playback rewrite for audio #2455, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add MediaSessionPlugin to playback rewrite #2457, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- adds "shuffle with" button if "always ask" is selected as preferred player\. #2459, by [@tim\-vk](https://github.com/tim\-vk)
- Add PlaybackManagerOptions for defaultRewindAmount and defaultFastForwardAmount #2473, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Update item delete confirm message string #2477, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Rewrite LibraryDreamService with Compose #2498, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Improve Android TV home screen recommendations #2515, by [@OctoNezd](https://github.com/OctoNezd)
- Remove playlist creation from NowPlaying screen #2523, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Update "Latest" media labels to be "Recently Added" #2549, by [@thornbill](https://github.com/thornbill)
- Always use getQuantityString\(\) for time units #2553, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix all string capitalization #2563, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Use plurals for amount of tracks in album / episodes in season #2571, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Explicitly use JDK 17 through gradle toolchains #2572, by [@DavidFair](https://github.com/DavidFair)
- Add blurHashPainter composable #2585, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Remove watched indicator from empty series #2589, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add media manager implementation using new playback module #2593, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add androidx\.media2 MediaSessionService implementation #2594, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Update and specify full action versions #2600, by [@1337joe](https://github.com/1337joe)
- Add \(experimental\) reactive homepage #2644, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix BaseItemInfoRow not showing for persons #2659, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Rewrite GridButton with Compose #2670, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Remove BaseProfile #2691, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Migrate search to SDK to fix cancellation issues #2727, by [@polson](https://github.com/polson)
- Fix screensaver support in video player #2749, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Migrate ExoPlayer to StyledPlayerView #2754, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add playback order support to playback rewrite #2783, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Enable playback rewrite by default for music #2802, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Rewrite media session code with androidx\.media3 #2803, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add repeat mode to playback rewrite #2810, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Update unsupported server messages #2815, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add volume state to playback rewrite #2817, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Use fast forward & rewind durations from preferences #2825, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Replace Glide with Coil #2889, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Remove AccountManager support #2890, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add preference to disable device type detection notification #2892, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Improve codec info #2893, by [@mohd\-akram](https://github.com/mohd\-akram)
- Add music progress to screensaver #2922, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add animation labels for Compose #2927, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Simplify defaultRewindAmount and defaultFastForwardAmount options #2928, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Use rounded corner for item details image #2929, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add \(PlaySession\-\)identifier to MediaStream #2935, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Use Android 14 SDK \(API 34\) #2957, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add margin to media grid #2964, by [@DoggoOfSpeed](https://github.com/DoggoOfSpeed)
- Folder improvements #2979, by [@DoggoOfSpeed](https://github.com/DoggoOfSpeed)
- Alignment changes to audio player #2995, by [@DoggoOfSpeed](https://github.com/DoggoOfSpeed)
- Add transcoding and remuxing #3001, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Make aspect ratio setting use radio buttons #3012, by [@DoggoOfSpeed](https://github.com/DoggoOfSpeed)
- Grab background images no larger than the screen #3020, by [@DoggoOfSpeed](https://github.com/DoggoOfSpeed)
- Remove old user preference migrations \(before 0\.15\) #3041, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Migrate ExternalPlayerActivity \(mostly\) to SDK #3055, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Use vignette effect in screensaver to darken background image #3061, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add initial playback capability testing #3062, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Use layer\-list for app\_banner drawable #3085, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Enable support for Opus and FLAC in ExoPlayer #3108, by [@reacocard](https://github.com/reacocard)
- Support remux/transcode fallback in new music player #3113, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Implement queue management in RewriteMediaManager #3137, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)

## 💥 Crash fixes

- Revert "Update dependency com\.android\.tools:desugar\_jdk\_libs to v2\.0\.1" #2468, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix screensaver crash when loading an invalid backdrop image #2567, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix crash when player controls are auto\-hiding while closing the player #2848, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix application crash on startup #2884, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix crash when switching users while playing music #2923, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Always use child fragment manager instead of activity fragment manager #2939, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix crash when reusing actions in MyDetailsOverviewRowPresenter #2940, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Improve BaseRowItem refreshing \(fix crash after item deletion\) #2949, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix crash when same item appears in queue multiple times #2960, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix crash during media stream resolving for corrupt files #3056, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)

## 🔧 Bugfixes

- Ignore missing notification permission from Glide #2422, by [@teobaranga](https://github.com/teobaranga)
- Fix crash when scaleType is null in AsyncImage #2554, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix missing background in DreamContentLogo #2560, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix playback rewrite audio queue in UI #2649, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Disable MPEG transport stream container in ExoPlayer #2695, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix ExoPlayer extensions not used for music playback #2711, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Remove old screen locks #2746, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix HomeFragmentLatestRow using wrong import for runBlocking function #2747, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add fallback image to LeanbackChannelWorker #2753, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix app closing when media finishes playing #2756, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Use navigation\.goBack instead of activity finish in FullDetailsFragment #2757, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix view binding leak in fragments #2763, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix BackgroundService flickering #2773, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Disable custom key handling when using playback rewrite #2804, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix navigation reset called during back stack pop call #2869, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Add missing view action to StartupActivity #2879, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix video player background not black #2895, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix wrong lifecycle state in NextUpFragment #2938, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix UI corruption when resuming the FullDetailsFragment #2943, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix ExoPlayerBackend not triggering onMediaStreamEnd correctly #2961, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Remove recordings from search results #2973, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Synochronise music timers #2996, by [@DoggoOfSpeed](https://github.com/DoggoOfSpeed)
- Fix grid getting focus when empty #3003, by [@DoggoOfSpeed](https://github.com/DoggoOfSpeed)
- Fix user image being too cropped in #3005, by [@DoggoOfSpeed](https://github.com/DoggoOfSpeed)
- Fix sync issue in ItemRowAdapterHelper #3024, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix blurry banners and thumbnails with fillWidth #3036, by [@DoggoOfSpeed](https://github.com/DoggoOfSpeed)
- Fix ExternalPlayerActivity not using PreferenceStore\.reset #3040, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Restore session on MainActivity stop #3054, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix progress bar positioning in view\_row\_details #3058, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix embedded subtitles not loading when subtitle filter on library #3090, by [@horizon1250](https://github.com/horizon1250)
- Fix playback stop reported twice #3105, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix display mode not changing back when closing player #3115, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix GridButtonPresenter focus on old Android platforms #3133, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)

## 🔃 Refactoring

- Remove gradlew\-update\.yaml workflow #2445, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Split MediaManager into interface and implementation #2453, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Split MediaManager into MediaManager and VideoQueueManager #2458, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Remove selected photo item from MediaManager and do some more cleaning #2461, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Replace sharedViewModel with activityViewModel #2518, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Replace deprecated Bundle\.get with Bundle\.getParcelable #2519, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Replace deprecated ExistingPeriodicWorkPolicy\.REPLACE with ExistingPeriodicWorkPolicy\.UPDATE #2524, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Refactor CustomPlaybackTransportControlGlue to extract logic into actions #2569, by [@DavidFair](https://github.com/DavidFair)
- Remove some unused drawables/strings/layouts #2570, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Use ItemRowAdapterHelperKt\.setItems for NextUpQuery in ItemRowAdapter #2641, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Rewrite background service UI with compose #2652, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Enable nonTransitiveRClass #2664, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Enable nonFinalResIds #2665, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Migrate video player & next up screen to fragments #2745, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Inline validateAuthentication extension into MainActivity #2758, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Use context in ItemLauncher #2760, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Migrate some activity usages to context #2761, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Use view binding in MainActivity & StartupActivity #2762, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Remove left over search hint code #2772, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Replace XML menus with Kotlin #2779, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Convert queue service to core functionality #2801, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Remove unused ProcessLifecycleInitializer #2880, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Use data object in sealed interfaces #2920, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Replace flow\.collect usages with onEach & launchIn #2925, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Remove Thread\.sleep usages #3000, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fix compose deprecations #3035, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Fully migrate ItemLauncher to SDK #3045, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Migrate ItemListView to SDK #3046, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Migrate SuggestedMoviesFragment to SDK #3049, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)

## 💡 Everything else

- Add androidx\.compose\.compiler to libraries section in version catalog #2491, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Revert "Update aboutlibraries to v10\.6\.0" #2520, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- Enable Renovate dependency dashboard #2774, by [@nielsvanvelzen](https://github.com/nielsvanvelzen)

## Contributors

- [@nielsvanvelzen](https://github.com/nielsvanvelzen)
- [@DoggoOfSpeed](https://github.com/DoggoOfSpeed)
- [@DavidFair](https://github.com/DavidFair)
- [@jellyfin\-bot](https://github.com/jellyfin\-bot)
- [@Clubfan22](https://github.com/Clubfan22)
- [@MWisBest](https://github.com/MWisBest)
- [@teobaranga](https://github.com/teobaranga)
- [@tim\-vk](https://github.com/tim\-vk)
- [@OctoNezd](https://github.com/OctoNezd)
- [@thornbill](https://github.com/thornbill)
- [@1337joe](https://github.com/1337joe)
- [@polson](https://github.com/polson)
- [@mohd\-akram](https://github.com/mohd\-akram)
- [@horizon1250](https://github.com/horizon1250)
- [@reacocard](https://github.com/reacocard)

<hr /><em>This discussion was created from the release <a href='https://github.com/jellyfin/jellyfin-androidtv/releases/tag/v0.16.0'>v0.16.0</a>.</em>

**Comment by nylimited** [comment]
I have been reading your various threads with interest.  Thank you for all that work!

Just one quick question and I'll go back to my quiet reading mode ...  what happens to AndroidTV when Amazon switches to their (now called) Vega OS?

**Reply by thornbill** [reply]
I haven't seen any developer resources for it yet so 🤷‍♂️

**Reply by nielsvanvelzen** [reply]
This app will likely not work on their new OS because it doesn't use Android, so we will probably publish our web client to it.

**Reply by nylimited** [reply]
Thank you, both!
